client/app/components/QueryEditor.jsx

```diff
@@ -0,0 +1,240 @@
+      keywords: [], // eslint-disable-line react/no-unused-state
+      autocompleteQuery: true,
+    };
+    langTools.addCompleter({
+      getCompletions: (state, session, pos, prefix, callback) => {
+        if (prefix.length === 0) {
+          callback(null, []);
+          return;
+        }
+        callback(null, this.state.keywords);
+      },
+    });
+
+    this.onLoad = (editor) => {
+      // Release Cmd/Ctrl+L to the browser
+      editor.commands.bindKey('Cmd+L', null);
+      editor.commands.bindKey('Ctrl+P', null);
+      editor.commands.bindKey('Ctrl+L', null);
+
+      // eslint-disable-next-line react/prop-types
+      this.props.QuerySnippet.query((snippets) => {
+        const snippetManager = snippetsModule.snippetManager;
+        const m = {
+          snippetText: '',
+        };
+        m.snippets = snippetManager.parseSnippetFile(m.snippetText);
+        snippets.forEach((snippet) => {
+          m.snippets.push(snippet.getSnippet());
+        });
+        snippetManager.register(m.snippets || [], m.scope);
+      });
+      editor.focus();
+      this.props.listenForResize(() => editor.resize());
+      this.props.listenForEditorCommand((e, command, ...args) => {
+        switch (command) {
+          case 'focus': {
+            editor.focus();
+            break;
+          }
+          case 'paste': {
+            const [text] = args;
+            editor.session.doc.replace(editor.selection.getRange(), text);
+            const range = editor.selection.getRange();
+            this.props.updateQuery(editor.session.getValue());
+            editor.selection.setRange(range);
+            break;
+          }
+          default:
+            break;
+        }
+      });
+    };
+
+    this.formatQuery = () => {
+      // eslint-disable-next-line react/prop-types
+      const format = this.props.Query.format;
+      format(this.props.dataSource.syntax || 'sql', this.props.queryText)
+        .then(this.props.updateQuery)
+        .catch(error => toastr.error(error));
+    };
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    if (!nextProps.schema) {
+      return { keywords: [], autocompleteQuery: false };
+    } else if (nextProps.schema !== prevState.schema) {
+      return {
+        schema: nextProps.schema,
+        keywords: buildKeywordsFromSchema(nextProps.schema),
+        autocompleteQuery: (nextProps.schema.reduce((totalLength, table) =>
+          totalLength + table.columns.length, 0) <= 5000),
+      };
+    }
+    return null;
+  }
+
+  render() {
+    // eslint-disable-next-line react/prop-types
+    const modKey = this.props.KeyboardShortcuts.modKey;
+    const parameterTooltip = <span>Add New Parameter (<i>{modKey} + P</i>)</span>;
+    const formatTooltip = <span>Format Query</span>;
+    const saveTooltip = <span>{modKey} + S</span>;
+    const executeTooltip = <span>{modKey} + Enter</span>;
+    const acTooltip = <span>Autocomplete</span>;
```

Comment:
Why not define it where we use it?

Reply:
Sounds good. Using the bootstrap style these were a good bit longer so I pulled them out of line.

Reply:
fixed